File: src/main/java/jm/app/algorithm/MovingAverageModel.java
package jm.app.algorithm;

import jm.app.algebra.Matrix;
import jm.app.optimize.BasicModel;

public class MovingAverageModel implements BasicModel {

    private int windowsSize = 3;

    public MovingAverageModel() {
    }

    public MovingAverageModel(int windowsSize) {
        this.windowsSize = windowsSize;
    }

    @Override
    public Matrix optimize(Matrix x, Matrix y) {
        // no need to implement this method
        return null;
    }

    @Override
    public Matrix calcValue(Matrix x, Matrix params) {
        int length = x.getRowNum();
        Matrix yHat = new Matrix(length, 1);
        for (int i = 0; i < length; i++) {
            if (i < windowsSize) {
                double sum = 0.0;
                for (int j = i; j >= 0; j--) {
                    sum += x.getValue(j, 0).doubleValue();
                }
                double val = sum / (i + 1);
                yHat.setValue(i, 0, val);
            } else {
                double sum = 0.0;
                for (int j = i; j >= i - windowsSize + 1; j--) {
                    sum += x.getValue(j, 0).doubleValue();
                }
                double val = sum / windowsSize;
                yHat.setValue(i, 0, val);
            }
        }
        return yHat;
    }

    public int getWindowsSize() {
        return windowsSize;
    }

    public void setWindowsSize(int windowsSize) {
        this.windowsSize = windowsSize;
    }
}

File: py/algorithm/moving_average_model.py
default_windows_size = 3

def optimize(x, y):
    # no need to optimize the parameters
    params = []
    return params


def calc_value(x, params):
    y_hat = []
    for i in range(len(x)):
        if (i + 1 < default_windows_size):
            sum = 0.0
            for j in range(i + 1):
                sum = sum + x[j]
            val = sum / (i + 1)
            y_hat.append(val)
        else:
            sum = 0.0
            for j in range(i - default_windows_size + 1, i):
                sum = sum + x[j]
            val = sum / default_windows_size
            y_hat.append(val)
    return y_hat